[SPARK-12711][ML] ML StopWordsRemover does not protect itself from column name duplication

grzegorz-chilkiewicz opened the pull request:
Fixes problem and verifies fix by test suite.
Also - adds optional parameter: nullable (Boolean) to: SchemaUtils.appendColumn
and deduplicates SchemaUtils.appendColumn functions.

grzegorz-chilkiewicz committed:
[SPARK-12711] ML StopWordsRemover does not protect itself from column name duplication

grzegorz-chilkiewicz committed:
[SPARK-12711] ML StopWordsRemover does not protect itself from column name duplication

grzegorz-chilkiewicz committed:
Merge branch 'master' of https://github.com/grzegorz-chilkiewicz/spark into HEAD

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #2377 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2377/consoleFull)** for PR 10741 at commit [`ca9a852`](https://github.com/apache/spark/commit/ca9a8523fa81dce0260cbfd88585b90d06876a8f).

jkbradley commented:
Could you please add tag "[ML]" to the PR title?

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/StopWordsRemoverSuite.scala:
```diff
@@ -89,4 +89,22 @@ class StopWordsRemoverSuite
   }
+
+  test("StopWordsRemover output column already exists") {
+    val outpuCol = "expected"
```
typo: "outputCol"

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/StopWordsRemoverSuite.scala:
```diff
@@ -89,4 +89,22 @@ class StopWordsRemoverSuite
+    val remover = new StopWordsRemover()
+      .setInputCol("raw")
+      .setOutputCol(outpuCol)
+      .setCaseSensitive(true)
```
not needed for this test

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/StopWordsRemoverSuite.scala:
```diff
@@ -89,4 +89,22 @@ class StopWordsRemoverSuite
+    val thrown = intercept[IllegalArgumentException] {
+      testStopWordsRemover(remover, dataSet)
+    }
+    assert(thrown.getClass === classOf[IllegalArgumentException])
```
This is already checked in line 104.

grzegorz-chilkiewicz replied:
I tried to take pattern from:
github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/FailureSuite.scala#L100-L106

I think this check theoretically could be useful:
https://docs.oracle.com/javase/7/docs/api/java/lang/IllegalArgumentException.html
is non-abstract class and have subclasses,
intercept can catch e.g. PatternSyntaxException, which is not what we should treat as a test success.

But of course, PatternSyntaxException is very unlikely to have message "requirement failed: Column ${outpuCol} already exists."

Am I correct?
Maybe "assert(thrown.getClass === classOf[IllegalArgumentException])" line should stay?

jkbradley replied:
I see; I bet you're correct.  Don't worry about it for now, though.

jkbradley commented:
Also, please add a (short) PR description (in your first PR comment) since that will become part of the commit message.

SparkQA commented:
**[Test build #2377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2377/consoleFull)** for PR 10741 at commit [`ca9a852`](https://github.com/apache/spark/commit/ca9a8523fa81dce0260cbfd88585b90d06876a8f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

grzegorz-chilkiewicz committed:
[SPARK-12711][ML] ML StopWordsRemover does not protect itself from column name duplication

Fixes problem and verifies fix by test suite.
Also - adds optional parameter nullable (Boolean) to: SchemaUtils.appendColumn
and deduplicates SchemaUtils.appendColumn functions.

grzegorz-chilkiewicz commented:
Is everything ok with this PR?
Did I miss something?
Can we merge it with apache/spark ?

jkbradley commented:
LGTM
I'll run tests before merging

SparkQA commented:
**[Test build #2461 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2461/consoleFull)** for PR 10741 at commit [`37af391`](https://github.com/apache/spark/commit/37af3910430203793eddbc70988480ba33bfeef7).

SparkQA commented:
**[Test build #2461 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2461/consoleFull)** for PR 10741 at commit [`37af391`](https://github.com/apache/spark/commit/37af3910430203793eddbc70988480ba33bfeef7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/StopWordsRemoverSuite.scala:
```diff
@@ -89,4 +89,20 @@ class StopWordsRemoverSuite
+    val remover = new StopWordsRemover()
+      .setInputCol("raw")
+      .setOutputCol(outputCol)
+    val dataSet = sqlContext.createDataFrame(Seq(
```
Just copy one of the datasets from an above test.  That should fix the error.

grzegorz-chilkiewicz replied:
I missed out that second column in dataSet was totally empty... - and that was the problem...
I do not want to make that example too complicated, because this test does not even check correctness of execution result
I'm sorry for problems

grzegorz-chilkiewicz committed:
Fix empty DataFrame column error

marmbrus commented:
ok to test

SparkQA commented:
**[Test build #50573 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50573/consoleFull)** for PR 10741 at commit [`49fd362`](https://github.com/apache/spark/commit/49fd362d5f9426c74427fa089de750c3c4e67a2d).

jkbradley commented:
LGTM pending tests

SparkQA commented:
**[Test build #50573 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50573/consoleFull)** for PR 10741 at commit [`49fd362`](https://github.com/apache/spark/commit/49fd362d5f9426c74427fa089de750c3c4e67a2d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50573/
Test PASSed.

jkbradley commented:
Merging with master and branch-1.6
Thanks for the PR!
